core/trino-main/src/main/java/io/trino/operator/output/RleAwarePositionsAppender.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.output;
+
+import io.trino.spi.block.Block;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import io.trino.type.BlockTypeOperators.BlockPositionEqual;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+public class RleAwarePositionsAppender
+        implements PositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(RleAwarePositionsAppender.class).instanceSize();
+    private static final int EMPTY_STATE_MARKER = 0;
+    private static final int FLAT_STATE_MARKER = -1;
+
+    private final BlockPositionEqual equalOperator;
+    private final PositionsAppender flat;
+
+    @Nullable
+    private Block rleValue;
+
+    // FLAT_STATE_MARKER means flat state, EMPTY_STATE_MARKER means initial empty state, positive means rle state and the current rle position count.
+    private int rlePositionCount = EMPTY_STATE_MARKER;
+
+    public RleAwarePositionsAppender(BlockPositionEqual equalOperator, PositionsAppender flat)
+    {
+        this.flat = requireNonNull(flat, "appender is null");
+        this.equalOperator = equalOperator;
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block source)
+    {
```

Comment:
check state that `source` is not RLE. `PositionsAppender` doesn't guarantee that, comment is needed.

Reply:
added. PTAL